lib/evaluate/integer.h

```diff
@@ -0,0 +1,853 @@
+// types &/or partial fields in the parts.  The radix (i.e., the number
+// of possible values in a part), however, must be a power of two; this
+// template class is not generalized to enable, say, decimal arithmetic.
+// Member functions that correspond to Fortran intrinsic functions are
+// named accordingly so that they can be referenced easily in the
+// language standard.
+template<int BITS, bool IS_LITTLE_ENDIAN = IsHostLittleEndian,
+    int PARTBITS =
+        BITS<32 ? BITS : 32, typename PART = HostUnsignedInt<PARTBITS>,
+            typename BIGPART = HostUnsignedInt<PARTBITS * 2>> class Integer {
+public:
+  static constexpr int bits{BITS};
+  static constexpr int partBits{PARTBITS};
+  using Part = PART;
+  using BigPart = BIGPART;
+  static_assert(CHAR_BIT * sizeof(BigPart) >= 2 * partBits);
+  static constexpr bool littleEndian{IS_LITTLE_ENDIAN};
+
+private:
+  static constexpr int maxPartBits{CHAR_BIT * sizeof(Part)};
+  static_assert(partBits > 0 && partBits <= maxPartBits);
+  static constexpr int extraPartBits{maxPartBits - partBits};
+  static constexpr int parts{(bits + partBits - 1) / partBits};
+  static_assert(parts >= 1);
+  static constexpr int extraTopPartBits{
+      extraPartBits + (parts * partBits) - bits};
+  static constexpr int topPartBits{maxPartBits - extraTopPartBits};
+  static_assert(topPartBits > 0 && topPartBits <= partBits);
+  static_assert((parts - 1) * partBits + topPartBits == bits);
+  static constexpr Part partMask{static_cast<Part>(~0) >> extraPartBits};
+  static constexpr Part topPartMask{static_cast<Part>(~0) >> extraTopPartBits};
+
+public:
+  struct ValueWithOverflow {
+    Integer value;
+    bool overflow;
+  };
+
+  struct ValueWithCarry {
+    Integer value;
+    bool carry;
+  };
+
+  struct Product {
+    Integer upper, lower;
+  };
+
+  struct QuotientWithRemainder {
+    Integer quotient, remainder;
+    bool divisionByZero, overflow;
+  };
+
+  // Constructors and value-generating static functions
+  constexpr Integer() { Clear(); }  // default constructor: zero
+  constexpr Integer(const Integer &) = default;
+  constexpr Integer(std::uint64_t n) {
+    for (int j{0}; j + 1 < parts; ++j) {
+      SetLEPart(j, n);
+      if constexpr (partBits < 64) {
+        n >>= partBits;
+      } else {
+        n = 0;
+      }
+    }
+    SetLEPart(parts - 1, n);
+  }
+  constexpr Integer(std::int64_t n) {
+    std::int64_t signExtension{-(n < 0)};
+    signExtension <<= partBits;
+    for (int j{0}; j + 1 < parts; ++j) {
+      SetLEPart(j, n);
+      if constexpr (partBits < 64) {
+        n = (n >> partBits) | signExtension;
+      } else {
+        n = signExtension;
+      }
+    }
+    SetLEPart(parts - 1, n);
+  }
+
+  constexpr Integer &operator=(const Integer &) = default;
+
+  // Left-justified mask (e.g., MASKL(1) has only its sign bit set)
+  static constexpr Integer MASKL(int places) {
+    if (places <= 0) {
+      return {};
```

Comment:
I find using `{}` for zero disconcerting, though maybe you get used to it.
What do you think about adding `static constexpr Integer Zero() { return {}; }` so it can be written as `Zero()`?

Reply:
I guess that I am used to reading `{}` as default construction.  How about `{0}`?

Reply:
Given that we can't do the same with Real (i.e. we can't use `{0.0}`), I think leave it as `{}`.